scripts/linking.py

```diff
@@ -1,86 +1,241 @@
+    tfidfs = tfidf_vectorizer.transform(mention_texts)
+    start_time = datetime.datetime.now()
+    neighbors = nmslis_knn_with_zero_vectors(tfidfs, k, ann_index)
+    end_time = datetime.datetime.now()
+    total_time = end_time - start_time
+    print(f'Finding neighbors took {total_time.total_seconds()} seconds')
+    neighbors_by_concept_ids = []
+    for n in neighbors:
+        if n is None:
+            n = []
+        predicted_umls_concept_ids = set([ann_concept_id_list[x] for x in n])
+        neighbors_by_concept_ids.append(predicted_umls_concept_ids)
+    return neighbors_by_concept_ids
+
+def create_load_tfidf_ann_index(ann_index_path: str, tfidf_vectorizer_path: str, umls_concept_list: List):
     """
-    umls_concept_dict_by_id = {}
-    umls_concept_dict_by_name = defaultdict(list)
-    with open(umls_path) as f:
-        print('Loading umls concepts ... ')
-        concepts = json.load(f)
-        print('Building indices ... ')
-        for concept in concepts:
-            # concept id -> concept object  (UMLS calls them concepts, MedMentions calls them entities)
-            umls_concept_dict_by_id[concept['concept_id']] = concept
-            # normalized entity name -> entity object  (useful for exact matching)
-            umls_concept_dict_by_name[normalize(concept['canonical_name'])].append(concept)
-    print(f'Number of umls concepts: {len(umls_concept_dict_by_id)}')
-    print(f'Number of unique names: {len(umls_concept_dict_by_name)}')
-    return umls_concept_dict_by_id, umls_concept_dict_by_name
-
-def main(medmentions_path: str, umls_path: str):
-
-    umls_concept_dict_by_id, umls_concept_dict_by_name = prepare_umls_indices(umls_path)
-
-    print('Reading corpus ... ')
+    Build or load tfidf vectorizer and ann index
+    """
+    uml_concept_ids = []
+    uml_concept_aliases = []
+    print('Collecting aliases ... ')
+    for i, concept in enumerate(umls_concept_list):
+        concept_id = concept['concept_id']
+        concept_aliases = concept['aliases'] + [concept['canonical_name']]
+
+        # TODO: remove the following line and rebuild index
+        concept_aliases = [a if len(a) >= 2 else f'{a}00' for a in concept_aliases]
+
+        uml_concept_ids.extend([concept_id] * len(concept_aliases))
+        uml_concept_aliases.extend(concept_aliases)
+
+        if i % 1000000 == 0 and i > 0:
+            print(f'Processed {i} or {len(umls_concept_list)} concepts')
+
+    uml_concept_ids = np.array(uml_concept_ids)
+    uml_concept_aliases = np.array(uml_concept_aliases)
+    assert len(uml_concept_ids) == len(uml_concept_aliases)
+
+    if not os.path.isfile(tfidf_vectorizer_path):
+        print(f'No tfidf vectorizer on {tfidf_vectorizer_path}')
+        print(f'Fitting tfidf vectorizer on {len(uml_concept_aliases)} aliases')
+        tfidf_vectorizer = TfidfVectorizer(analyzer='char', ngram_range=(3, 4))
+        tfidf_vectorizer.fit(uml_concept_aliases)
```

Comment:
I believe you could combine these steps using `fit_transform`, I'm not sure how much of a speed boost you get - i'd imagine possibly substantial as here you are iterating twice over the collection.

Reply:
True, but this is only relevant at the time of fitting the `TfidfVectorizer` (will save less than 5 minutes), but won't speed up prediction or loading to memory.